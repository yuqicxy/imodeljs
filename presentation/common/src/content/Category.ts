/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module Content
 */

/**
 * A data structure that describes a [[Field]] category.
 * @public
 */
export interface CategoryDescription {
  /** Unique name */
  name: string;
  /** Display label */
  label: string;
  /** Extensive description */
  description: string;
  /** Priority. Categories with higher priority should appear higher in the UI */
  priority: number;
  /** Should this category be auto-expanded when it's displayed in the UI */
  expand: boolean;
}
