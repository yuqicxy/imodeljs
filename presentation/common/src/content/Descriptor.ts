/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module Content
 */

import {
  ClassInfo, ClassInfoJSON,
  RelatedClassInfo, RelationshipPath, RelationshipPathJSON, RelatedClassInfoJSON,
} from "../EC";
import { Field, FieldJSON, getFieldByName } from "./Fields";

/**
 * Data structure that describes an ECClass in content [[Descriptor]].
 * @public
 */
export interface SelectClassInfo {
  /** Information about the ECClass */
  selectClassInfo: ClassInfo;
  /** Is the class handled polymorphically */
  isSelectPolymorphic: boolean;
  /** Relationship path to the [Primary class]($docs/learning/presentation/Content/Terminology#primary-class) */
  pathToPrimaryClass: RelationshipPath;
  /** Relationship paths to [Related property]($docs/learning/presentation/Content/Terminology#related-properties) classes */
  relatedPropertyPaths: RelationshipPath[];
  /** Relationship paths to navigation property classes */
  navigationPropertyClasses: RelatedClassInfo[];
  /** Relationship paths to [Related instance]($docs/learning/presentation/Content/Terminology#related-instance) classes */
  relatedInstanceClasses: RelatedClassInfo[];
}

/**
 * Serialized [[SelectClassInfo]]
 * @internal
 */
export interface SelectClassInfoJSON {
  selectClassInfo: ClassInfoJSON;
  isSelectPolymorphic: boolean;
  pathToPrimaryClass: RelationshipPathJSON;
  relatedPropertyPaths: RelationshipPathJSON[];
  navigationPropertyClasses: RelatedClassInfoJSON[];
  relatedInstanceClasses: RelatedClassInfoJSON[];
}

const selectClassInfoFromJSON = (json: SelectClassInfoJSON): SelectClassInfo => {
  return {
    ...json,
    selectClassInfo: ClassInfo.fromJSON(json.selectClassInfo),
    pathToPrimaryClass: json.pathToPrimaryClass.map((p) => RelatedClassInfo.fromJSON(p)),
    relatedPropertyPaths: json.relatedPropertyPaths.map((rp) => (rp.map((p) => RelatedClassInfo.fromJSON(p)))),
    navigationPropertyClasses: json.navigationPropertyClasses.map((p) => RelatedClassInfo.fromJSON(p)),
    relatedInstanceClasses: json.relatedInstanceClasses.map((p) => RelatedClassInfo.fromJSON(p)),
  };
};

/**
 * Flags that control content format.
 * @public
 */
export enum ContentFlags {
  /** Each content record only has [[InstanceKey]] and no data */
  KeysOnly = 1 << 0,

  /** Each content record additionally has an image id */
  ShowImages = 1 << 1,

  /** Each content record additionally has a display label */
  ShowLabels = 1 << 2,

  /** All content records are merged into a single record (see [Merging values]($docs/learning/terminology#value-merging)) */
  MergeResults = 1 << 3,

  /** Content has only distinct values */
  DistinctValues = 1 << 4,

  /** Doesn't create property or calculated fields. Can be used in conjunction with [[ShowLabels]]. */
  NoFields = 1 << 5,
}

/**
 * Data sorting direction
 * @public
 */
export enum SortDirection {
  Ascending,
  Descending,
}

/**
 * Data structure that contains selection information. Used
 * for cases when requesting content after a selection change.
 *
 * @public
 */
export interface SelectionInfo {
  /** Name of selection provider which cause the selection change */
  providerName: string;
  /** Level of selection that changed */
  level?: number;
}

/**
 * Serialized [[Descriptor]] JSON representation.
 * @internal
 */
export interface DescriptorJSON {
  connectionId: string;
  inputKeysHash: string;
  contentOptions: any;
  selectionInfo?: SelectionInfo;
  displayType: string;
  selectClasses: SelectClassInfoJSON[];
  fields: FieldJSON[];
  sortingFieldName?: string;
  sortDirection?: SortDirection;
  contentFlags: number;
  filterExpression?: string;
}

/**
 * Descriptor overrides that can be used to customize
 * content.
 *
 * @public
 */
export interface DescriptorOverrides {
  /** Content display type. Can be accessed in presentation rules and used to modify content in various ways */
  displayType: string;
  /** Names of fields which should be excluded from content */
  hiddenFieldNames: string[];
  /** Content flags used for content customization. See [[ContentFlags]] */
  contentFlags: number;
  /** Name of the sorting field */
  sortingFieldName?: string;
  /** Sort direction. Defaults to [[SortDirection.Ascending]] */
  sortDirection?: SortDirection;
  /** [ECExpression]($docs/learning/presentation/ECExpressions.md) for filtering content */
  filterExpression?: string;
}

/**
 * Descriptor properties
 * @public
 */
export interface DescriptorSource {
  /** Selection info used to create the descriptor */
  selectionInfo?: SelectionInfo;
  /** Display type used to create the descriptor */
  displayType: string;
  /** A list of classes that will be selected from when creating content with this descriptor */
  selectClasses: SelectClassInfo[];
  /** A list of fields contained in the descriptor */
  fields: Field[];
  /** [[ContentFlags]] used to create the descriptor */
  contentFlags: number;
  /** Field used to sort the content */
  sortingField?: Field;
  /** Sorting direction */
  sortDirection?: SortDirection;
  /** Content filtering [ECExpression]($docs/learning/presentation/ECExpressions) */
  filterExpression?: string;
}

/**
 * Data structure that describes content: fields, sorting, filtering, format, etc.
 * Descriptor may be changed to control how content is created.
 *
 * @public
 */
export class Descriptor implements DescriptorSource {
  /** Id of the connection used to create the descriptor */
  public connectionId!: string;
  /** Hash of the input keys used to create the descriptor */
  public inputKeysHash!: string;
  /** Extended options used to create the descriptor */
  public contentOptions: any;
  /** Selection info used to create the descriptor */
  public selectionInfo?: SelectionInfo;
  /** Display type used to create the descriptor */
  public displayType!: string;
  /** A list of classes that will be selected from when creating content with this descriptor */
  public selectClasses!: SelectClassInfo[];
  /** A list of fields contained in the descriptor */
  public fields!: Field[];
  /** [[ContentFlags]] used to create the descriptor */
  public contentFlags!: number;
  /** Field used to sort the content */
  public sortingField?: Field;
  /** Sorting direction */
  public sortDirection?: SortDirection;
  /** Content filtering [ECExpression]($docs/learning/presentation/ECExpressions) */
  public filterExpression?: string;

  /** Construct a new Descriptor using a `DescriptorSource` */
  public constructor(source: DescriptorSource) {
    Object.assign(this, source, {
      selectClasses: [...source.selectClasses],
      fields: [...source.fields],
    });
  }

  /** @internal */
  public toJSON(): DescriptorJSON {
    return Object.assign({}, this, {
      fields: this.fields.map((field: Field) => field.toJSON()),
    });
  }

  /**
   * Deserialize Descriptor from JSON
   * @param json JSON or JSON serialized to string to deserialize from
   * @returns Deserialized descriptor or undefined if deserialization failed
   *
   * @internal
   */
  public static fromJSON(json: DescriptorJSON | string | undefined): Descriptor | undefined {
    if (!json)
      return undefined;
    if (typeof json === "string")
      return JSON.parse(json, Descriptor.reviver);
    const descriptor = Object.create(Descriptor.prototype);
    return Object.assign(descriptor, json, {
      fields: json.fields.map((fieldJson: FieldJSON) => {
        const field = Field.fromJSON(fieldJson);
        if (field)
          field.rebuildParentship();
        return field;
      }).filter((field) => (undefined !== field)),
      selectClasses: json.selectClasses.map((selectClass: SelectClassInfoJSON) => selectClassInfoFromJSON(selectClass)),
    } as Partial<Descriptor>);
  }

  /**
   * Reviver function that can be used as a second argument for
   * `JSON.parse` method when parsing Content objects.
   *
   * @internal
   */
  public static reviver(key: string, value: any): any {
    return key === "" ? Descriptor.fromJSON(value) : value;
  }

  /**
   * Get field by its name
   * @param name Name of the field to find
   * @param recurse Recurse into nested fields
   */
  public getFieldByName(name: string, recurse?: boolean): Field | undefined {
    return getFieldByName(this.fields, name, recurse);
  }

  /** @internal */
  public createDescriptorOverrides(): DescriptorOverrides {
    return {
      displayType: this.displayType,
      hiddenFieldNames: [],
      sortingFieldName: this.sortingField ? this.sortingField.name : undefined,
      sortDirection: this.sortDirection,
      contentFlags: this.contentFlags,
      filterExpression: this.filterExpression,
    };
  }

  /** @internal */
  public createStrippedDescriptor(): Descriptor {
    const stripped = Object.create(Descriptor.prototype);
    return Object.assign(stripped, this, {
      fields: [],
      selectClasses: [],
    });
  }
}
