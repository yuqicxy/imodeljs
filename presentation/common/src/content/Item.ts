/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module Content
 */

import {
  ClassInfo, ClassInfoJSON,
  InstanceKey, InstanceKeyJSON,
} from "../EC";
import { ValuesDictionary } from "../Utils";
import {
  Value, DisplayValue,
  DisplayValueJSON, ValueJSON,
  ValuesMapJSON, DisplayValuesMapJSON,
} from "./Value";
import { LabelDefinition, LabelDefinitionJSON } from "../LabelDefinition";

/**
 * Serialized [[Item]] JSON representation.
 * @internal
 */
export interface ItemJSON {
  primaryKeys: InstanceKeyJSON[];
  /** @deprecated use labelDefinition instead */
  label?: string;
  labelDefinition: LabelDefinitionJSON;
  imageId: string;
  classInfo?: ClassInfoJSON;
  values: ValuesDictionary<ValueJSON>;
  displayValues: ValuesDictionary<DisplayValueJSON>;
  mergedFieldNames: string[];
  extendedData?: { [key: string]: any };
}

/**
 * A data structure that represents a single content record.
 * @public
 */
export class Item {
  /** Keys of instances whose data is contained in this item */
  public primaryKeys: InstanceKey[];
  /** Display label of the item
   * @deprecated use 'labelDefinition' instead
   */
  public label: string;
  /** Definition of item display label */
  public labelDefinition: LabelDefinition;
  /** ID of the image associated with this item */
  public imageId: string;
  /** For cases when item consists only of same class instances, information about the ECClass */
  public classInfo?: ClassInfo;
  /** Raw values dictionary */
  public values: ValuesDictionary<Value>;
  /** Display values dictionary */
  public displayValues: ValuesDictionary<DisplayValue>;
  /** List of field names whose values are merged (see [Merging values]($docs/learning/presentation/Content/Terminology#value-merging)) */
  public mergedFieldNames: string[];
  /** Extended data injected into this content item */
  public extendedData?: { [key: string]: any };

  /**
   * Creates an instance of Item.
   * @param primaryKeys Keys of instances whose data is contained in this item
   * @param label Display label of the item
   * @param imageId ID of the image associated with this item
   * @param classInfo For cases when item consists only of same class instances, information about the ECClass
   * @param values Raw values dictionary
   * @param displayValues Display values dictionary
   * @param mergedFieldNames List of field names whose values are merged (see [Merging values]($docs/learning/presentation/Content/Terminology#value-merging))
   * @param extendedData Extended data injected into this content item
   */
  public constructor(primaryKeys: InstanceKey[], label: string | LabelDefinition, imageId: string, classInfo: ClassInfo | undefined,
    values: ValuesDictionary<Value>, displayValues: ValuesDictionary<DisplayValue>, mergedFieldNames: string[], extendedData?: { [key: string]: any }) {
    this.primaryKeys = primaryKeys;
    this.imageId = imageId;
    this.classInfo = classInfo;
    this.values = values;
    this.displayValues = displayValues;
    this.mergedFieldNames = mergedFieldNames;
    this.extendedData = extendedData;
    this.label = typeof label === "string" ? label : label.displayValue;
    this.labelDefinition = typeof label === "string" ? LabelDefinition.fromLabelString(label) : label;
  }

  /**
   * Is value of field with the specified name merged in this record.
   */
  public isFieldMerged(fieldName: string): boolean {
    return -1 !== this.mergedFieldNames.indexOf(fieldName);
  }

  /** @internal */
  public toJSON(): ItemJSON {
    const { labelDefinition, ...content } = this;
    return Object.assign({}, content, {
      classInfo: this.classInfo ? ClassInfo.toJSON(this.classInfo) : undefined,
      values: Value.toJSON(this.values) as ValuesMapJSON,
      displayValues: DisplayValue.toJSON(this.displayValues) as DisplayValuesMapJSON,
      labelDefinition: LabelDefinition.toJSON(labelDefinition),
      label: labelDefinition.displayValue,
    });
  }

  /**
   * Deserialize Item from JSON
   * @param json JSON or JSON serialized to string to deserialize from
   * @returns Deserialized item or undefined if deserialization failed
   *
   * @internal
   */
  public static fromJSON(json: ItemJSON | string | undefined): Item | undefined {
    if (!json)
      return undefined;
    if (typeof json === "string")
      return JSON.parse(json, Item.reviver);
    const item = Object.create(Item.prototype);
    return Object.assign(item, json, {
      primaryKeys: json.primaryKeys.map((pk) => InstanceKey.fromJSON(pk)),
      classInfo: json.classInfo ? ClassInfo.fromJSON(json.classInfo) : undefined,
      values: Value.fromJSON(json.values),
      displayValues: DisplayValue.fromJSON(json.displayValues),
      labelDefinition: LabelDefinition.fromJSON(json.labelDefinition),
      label: json.labelDefinition.displayValue,
    } as Partial<Item>);
  }

  /**
   * Reviver function that can be used as a second argument for
   * `JSON.parse` method when parsing Item objects.
   *
   * @internal
   */
  public static reviver(key: string, value: any): any {
    return key === "" ? Item.fromJSON(value) : value;
  }
}
