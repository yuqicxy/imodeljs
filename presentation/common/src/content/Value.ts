/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module Content
 */

import { InstanceKey, InstanceKeyJSON } from "../EC";
import { ValuesDictionary } from "../Utils";

/**
 * Raw value type
 * @public
 */
export type Value = string | number | boolean | undefined | ValuesMap | ValuesArray | NestedContentValue[];
/** @public */
export namespace Value {
  /** Is the value a primitive */
  export function isPrimitive(value: Value): value is string | number | boolean | undefined { return isPrimitiveValue(value); }
  /** Is the value an array */
  export function isArray(value: Value): value is ValuesArray { return isArrayValue(value); }
  /** Is the value a map / struct */
  export function isMap(value: Value): value is ValuesMap { return isMapValue(value); }
  /** Is the value a nested content value */
  export function isNestedContent(value: Value): value is NestedContentValue[] { return isNestedContentValue(value); }
  /** @internal */
  export function fromJSON(json: ValueJSON): Value {
    if (json === null)
      return undefined;
    if (isNestedContentValue(json))
      return json.map(NestedContentValue.fromJSON);
    if (isArrayValue(json))
      return valuesArrayFromJSON(json);
    if (isMapValue(json))
      return valuesMapFromJSON(json);
    return json;
  }
  /** @internal */
  export function toJSON(value: Value): ValueJSON {
    if (undefined === value)
      return null;
    if (isNestedContentValue(value))
      return value.map(NestedContentValue.toJSON);
    if (isArrayValue(value))
      return valuesArrayToJSON(value);
    if (isMapValue(value))
      return valuesMapToJSON(value);
    return value;
  }
}
/**
 * A map of raw values
 * @public
 */
export interface ValuesMap extends ValuesDictionary<Value> { }
/**
 * An array of raw values
 * @public
 */
export interface ValuesArray extends Array<Value> { }

/**
 * Display value type.
 * @public
 */
export type DisplayValue = string | undefined | DisplayValuesMap | DisplayValuesArray;
/** @public */
export namespace DisplayValue {
  /** Is the value a primitive */
  export function isPrimitive(value: DisplayValue): value is string | undefined { return isPrimitiveValue(value); }
  /** Is the value an array */
  export function isArray(value: DisplayValue): value is DisplayValuesArray { return isArrayValue(value); }
  /** Is the value a map / struct */
  export function isMap(value: DisplayValue): value is DisplayValuesMap { return isMapValue(value); }
  /** @internal */
  export function fromJSON(json: DisplayValueJSON): DisplayValue {
    if (json === null)
      return undefined;
    if (isArrayValue(json))
      return displayValuesArrayFromJSON(json);
    if (isMapValue(json))
      return displayValuesMapFromJSON(json);
    return json;
  }
  /** @internal */
  export function toJSON(value: DisplayValue): DisplayValueJSON {
    if (undefined === value)
      return null;
    if (isArrayValue(value))
      return displayValuesArrayToJSON(value);
    if (isMapValue(value))
      return displayValuesMapToJSON(value);
    return value;
  }
}
/**
 * A map of display values
 * @public
 */
export interface DisplayValuesMap extends ValuesDictionary<DisplayValue> { }
/**
 * An array of display values
 * @public
 */
export interface DisplayValuesArray extends Array<DisplayValue> { }

/**
 * Data structure that describes nested content value.
 * @public
 */
export interface NestedContentValue {
  /** Keys of instances whose content is contained in this value */
  primaryKeys: InstanceKey[];
  /** Content values map */
  values: ValuesDictionary<Value>;
  /** Content display values map */
  displayValues: ValuesDictionary<DisplayValue>;
  /** Names of fields whose values are merged */
  mergedFieldNames: string[];
}
/** @public */
export namespace NestedContentValue {
  /** @internal */
  export function toJSON(json: NestedContentValue): NestedContentValueJSON {
    return {
      primaryKeys: json.primaryKeys.map(InstanceKey.toJSON),
      values: valuesMapToJSON(json.values),
      displayValues: displayValuesMapToJSON(json.displayValues),
      mergedFieldNames: json.mergedFieldNames,
    };
  }
  /** @internal */
  export function fromJSON(json: NestedContentValueJSON): NestedContentValue {
    return {
      primaryKeys: json.primaryKeys.map(InstanceKey.fromJSON),
      values: valuesMapFromJSON(json.values),
      displayValues: displayValuesMapFromJSON(json.displayValues),
      mergedFieldNames: json.mergedFieldNames,
    };
  }
}

/** @internal */
export type ValueJSON = string | number | boolean | null | ValuesMapJSON | ValuesArrayJSON | NestedContentValueJSON[];
/** @internal */
export interface ValuesMapJSON extends ValuesDictionary<ValueJSON> { }
/** @internal */
export interface ValuesArrayJSON extends Array<ValueJSON> { }

/** @internal */
export type DisplayValueJSON = string | null | DisplayValuesMapJSON | DisplayValuesArrayJSON;
/** @internal */
export interface DisplayValuesMapJSON extends ValuesDictionary<DisplayValueJSON> { }
/** @internal */
export interface DisplayValuesArrayJSON extends Array<DisplayValueJSON> { }

/**
 * Serialized [[NestedContentValue]] JSON representation.
 * @internal
 */
export interface NestedContentValueJSON {
  primaryKeys: InstanceKeyJSON[];
  values: ValuesDictionary<ValueJSON>;
  displayValues: ValuesDictionary<DisplayValueJSON>;
  mergedFieldNames: string[];
}

function isNestedContentValue(v: Value | ValueJSON): v is NestedContentValue[] | NestedContentValueJSON[] {
  return (v !== undefined) && Array.isArray(v)
    && ((v.length === 0)
      || (v[0] as NestedContentValue).primaryKeys !== undefined
      && (v[0] as NestedContentValue).values !== undefined
      && (v[0] as NestedContentValue).displayValues !== undefined
      && (v[0] as NestedContentValue).mergedFieldNames !== undefined);
}
function isArrayValue(v: Value | ValueJSON | DisplayValue | DisplayValueJSON): v is ValuesArray | ValuesArrayJSON | DisplayValuesArray | DisplayValuesArrayJSON {
  // note: we don't guarantee by 100% that v is ValuesArray | DisplayValuesArray, but merely make compiler happy.
  // we have other means to determine the type of value.
  return (v !== undefined) && Array.isArray(v);
}
function isMapValue(v: Value | ValueJSON | DisplayValue | DisplayValueJSON): v is ValuesMap | ValuesMapJSON | DisplayValuesMap | DisplayValuesMapJSON {
  return (v !== undefined) && (typeof v === "object") && !Array.isArray(v);
}
function isPrimitiveValue(v: Value | DisplayValue): v is string | number | boolean | undefined {
  return !isArrayValue(v) && !isMapValue(v);
}

function valuesArrayFromJSON(json: ValuesArrayJSON): ValuesArray {
  return json.map(Value.fromJSON);
}
function valuesArrayToJSON(values: ValuesArray): ValuesArrayJSON {
  return values.map(Value.toJSON);
}
function valuesMapFromJSON(json: ValuesMapJSON): ValuesMap {
  const map: ValuesMap = {};
  for (const key in json) {
    /* istanbul ignore else */
    if (json.hasOwnProperty(key)) {
      map[key] = Value.fromJSON(json[key]);
    }
  }
  return map;
}
function valuesMapToJSON(values: ValuesMap): ValuesMapJSON {
  const map: ValuesMapJSON = {};
  for (const key in values) {
    /* istanbul ignore else */
    if (values.hasOwnProperty(key)) {
      map[key] = Value.toJSON(values[key]);
    }
  }
  return map;
}

function displayValuesArrayFromJSON(json: DisplayValuesArrayJSON): DisplayValuesArray {
  return json.map(DisplayValue.fromJSON);
}
function displayValuesArrayToJSON(values: DisplayValuesArray): DisplayValuesArrayJSON {
  return values.map(DisplayValue.toJSON);
}
function displayValuesMapFromJSON(json: DisplayValuesMapJSON): DisplayValuesMap {
  const map: DisplayValuesMap = {};
  for (const key in json) {
    /* istanbul ignore else */
    if (json.hasOwnProperty(key)) {
      map[key] = DisplayValue.fromJSON(json[key]);
    }
  }
  return map;
}
function displayValuesMapToJSON(values: DisplayValuesMap): DisplayValuesMapJSON {
  const map: DisplayValuesMapJSON = {};
  for (const key in values) {
    /* istanbul ignore else */
    if (values.hasOwnProperty(key)) {
      map[key] = DisplayValue.toJSON(values[key]);
    }
  }
  return map;
}
