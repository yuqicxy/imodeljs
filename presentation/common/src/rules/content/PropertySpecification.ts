/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module PresentationRules
 */

import { PropertyEditorSpecification } from "./modifiers/PropertyEditorsSpecification";

/**
 * A container structure for possible property overrides
 * @beta
 */
export interface PropertyOverrides {
  /** Priority of the specified overrides */
  overridesPriority?: number;
  /** Label override. May be [localized]($docs/learning/presentation/Localization.md). */
  labelOverride?: string;
  /** ID of a category specified through `PropertyCategorySpecification` in this scope. */
  categoryId?: string;
  /** Display override. `true` to force display, `false` to force hide, `undefined` to use default. */
  isDisplayed?: boolean;
  /** Custom property editor specification */
  editor?: PropertyEditorSpecification;
}

/**
 * Specification of an ECProperty and its overrides
 * @beta
 */
export interface PropertySpecification extends PropertyOverrides {
  /** Name of the ECProperty */
  name: string;
}
