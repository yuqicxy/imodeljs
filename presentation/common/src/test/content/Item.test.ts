/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
import { expect } from "chai";
import * as faker from "faker";
import { createRandomECClassInfoJSON, createRandomECInstanceKeyJSON, createRandomLabelDefinitionJSON, createRandomLabelDefinition } from "../_helpers/random";
import { Item, ItemJSON } from "../../content/Item";
import { NestedContentValueJSON } from "../../content/Value";

describe("Item", () => {

  describe("constructor", () => {

    it("creates valid item with label", () => {
      const item = new Item([], faker.random.word(), faker.random.uuid(),
        undefined, { key: faker.random.word() }, { key: faker.random.word() }, []);
      expect(item).to.matchSnapshot();
    });

    it("creates valid item with label definition", () => {
      const item = new Item([], createRandomLabelDefinition(), faker.random.uuid(),
        undefined, { key: faker.random.word() }, { key: faker.random.word() }, []);
      expect(item).to.matchSnapshot();
    });

  });

  describe("fromJSON", () => {

    let testItemJSON!: ItemJSON;
    beforeEach(() => {
      testItemJSON = {
        primaryKeys: [],
        labelDefinition: createRandomLabelDefinitionJSON(),
        imageId: faker.random.uuid(),
        classInfo: createRandomECClassInfoJSON(),
        values: {
          key1: faker.random.number(),
          key2: faker.random.words(),
        },
        displayValues: {
          key1: faker.random.words(),
          key2: faker.random.words(),
        },
        mergedFieldNames: ["key1"],
      };
    });

    it("creates valid Item from valid JSON", () => {
      const item = Item.fromJSON(testItemJSON);
      expect(item).to.matchSnapshot();
    });

    it("creates valid Item from valid serialized JSON", () => {
      const item = Item.fromJSON(JSON.stringify(testItemJSON));
      expect(item).to.matchSnapshot();
    });

    it("creates valid Item from valid JSON without classInfo", () => {
      const item = Item.fromJSON({ ...testItemJSON, classInfo: undefined });
      expect(item).to.matchSnapshot();
    });

    it("creates valid Item with null values", () => {
      const item = Item.fromJSON({
        ...testItemJSON,
        values: { key1: null },
        displayValues: { key1: null },
      });
      expect(item).to.matchSnapshot();
    });

    it("creates valid Item with nested content values", () => {
      const nestedContentValueJSON: NestedContentValueJSON = {
        primaryKeys: [createRandomECInstanceKeyJSON()],
        values: { nested: null },
        displayValues: { nested: "" },
        mergedFieldNames: [faker.random.word()],
      };
      const item = Item.fromJSON({
        ...testItemJSON,
        displayValues: { key1: null },
        values: {
          key1: [nestedContentValueJSON],
        },
      });
      expect(item).to.matchSnapshot();
    });

    it("returns undefined for undefined JSON", () => {
      const item = Item.fromJSON(undefined);
      expect(item).to.be.undefined;
    });

  });

  describe("isFieldMerged", () => {

    it("returns false for unmerged field", () => {
      const item = new Item([], faker.random.word(), faker.random.uuid(),
        undefined, { key: faker.random.word() }, { key: faker.random.word() }, []);
      expect(item.isFieldMerged("key")).to.be.false;
    });

    it("returns true for merged field", () => {
      const item = new Item([], faker.random.word(), faker.random.uuid(),
        undefined, { key: faker.random.word() }, { key: faker.random.word() }, ["key"]);
      expect(item.isFieldMerged("key")).to.be.true;
    });

  });

});
