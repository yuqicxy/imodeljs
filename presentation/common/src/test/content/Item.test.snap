// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item constructor creates valid item with label 1`] = `
Object {
  "classInfo": undefined,
  "displayValues": Object {
    "key": "Kina",
  },
  "extendedData": undefined,
  "imageId": "70c6cab5-86c7-4d22-b010-6c0cf48d9662",
  "label": "pink",
  "labelDefinition": Object {
    "displayValue": "pink",
    "rawValue": "pink",
    "typeName": "string",
  },
  "mergedFieldNames": Array [],
  "primaryKeys": Array [],
  "values": Object {
    "key": "models",
  },
}
`;

exports[`Item constructor creates valid item with label definition 1`] = `
Object {
  "classInfo": undefined,
  "displayValues": Object {
    "key": "Keyboard",
  },
  "extendedData": undefined,
  "imageId": "0ea0e226-a243-44c2-83b1-e941fb2d30c8",
  "label": "Oregon",
  "labelDefinition": Object {
    "displayValue": "Oregon",
    "rawValue": "Toys",
    "typeName": "string",
  },
  "mergedFieldNames": Array [],
  "primaryKeys": Array [],
  "values": Object {
    "key": "Savings Account",
  },
}
`;

exports[`Item fromJSON creates valid Item from valid JSON 1`] = `
Object {
  "classInfo": Object {
    "id": "0x9c64000000847e",
    "label": "Brand",
    "name": "XSS",
  },
  "displayValues": Object {
    "key1": "Associate Savings Account",
    "key2": "technologies",
  },
  "imageId": "4c3cfc6d-ca16-4409-b743-4afabfe4b926",
  "label": "Rustic Soft Fish",
  "labelDefinition": Object {
    "displayValue": "Rustic Soft Fish",
    "rawValue": "override",
    "typeName": "string",
  },
  "mergedFieldNames": Array [
    "key1",
  ],
  "primaryKeys": Array [],
  "values": Object {
    "key1": 94837,
    "key2": "mesh collaborative",
  },
}
`;

exports[`Item fromJSON creates valid Item from valid JSON without classInfo 1`] = `
Object {
  "classInfo": undefined,
  "displayValues": Object {
    "key1": "Rubber one-to-one Barbados Dollar",
    "key2": "Quetzal context-sensitive",
  },
  "imageId": "64b46a85-7e4f-428a-98df-26ce9acb15ed",
  "label": "Washington",
  "labelDefinition": Object {
    "displayValue": "Washington",
    "rawValue": "Common",
    "typeName": "string",
  },
  "mergedFieldNames": Array [
    "key1",
  ],
  "primaryKeys": Array [],
  "values": Object {
    "key1": 56931,
    "key2": "Optimization Kazakhstan Personal Loan Account",
  },
}
`;

exports[`Item fromJSON creates valid Item from valid serialized JSON 1`] = `
Object {
  "classInfo": Object {
    "id": "0xf3eb0000001909",
    "label": "magnetic International",
    "name": "Consultant",
  },
  "displayValues": Object {
    "key1": "Mobility",
    "key2": "payment Future Legacy",
  },
  "imageId": "fc61ba21-2b4d-4c31-9124-d4b70a2f087a",
  "label": "Strategist",
  "labelDefinition": Object {
    "displayValue": "Strategist",
    "rawValue": "firewall",
    "typeName": "string",
  },
  "mergedFieldNames": Array [
    "key1",
  ],
  "primaryKeys": Array [],
  "values": Object {
    "key1": 7721,
    "key2": "Books whiteboard discrete",
  },
}
`;

exports[`Item fromJSON creates valid Item with nested content values 1`] = `
Object {
  "classInfo": Object {
    "id": "0x6e520000005776",
    "label": "invoice Sleek",
    "name": "bandwidth",
  },
  "displayValues": Object {
    "key1": null,
  },
  "imageId": "425d312e-81f7-4a06-a30a-1de9fba7ba14",
  "label": "bricks-and-clicks",
  "labelDefinition": Object {
    "displayValue": "bricks-and-clicks",
    "rawValue": "Mountains",
    "typeName": "string",
  },
  "mergedFieldNames": Array [
    "key1",
  ],
  "primaryKeys": Array [],
  "values": Object {
    "key1": Array [
      Object {
        "displayValues": Object {
          "nested": "",
        },
        "mergedFieldNames": Array [
          "override",
        ],
        "primaryKeys": Array [
          Object {
            "className": "SAS",
            "id": "0x110880000004de8",
          },
        ],
        "values": Object {
          "nested": null,
        },
      },
    ],
  },
}
`;

exports[`Item fromJSON creates valid Item with null values 1`] = `
Object {
  "classInfo": Object {
    "id": "0xbd4c0000006e24",
    "label": "Incredible Metal Chair",
    "name": "deposit",
  },
  "displayValues": Object {
    "key1": null,
  },
  "imageId": "26da4e9d-28ab-4e7b-8895-4da11390a707",
  "label": "impactful",
  "labelDefinition": Object {
    "displayValue": "impactful",
    "rawValue": "GB",
    "typeName": "string",
  },
  "mergedFieldNames": Array [
    "key1",
  ],
  "primaryKeys": Array [],
  "values": Object {
    "key1": null,
  },
}
`;

exports[`Item toJSON serializes item without label definition 1`] = `
Object {
  "classInfo": undefined,
  "displayValues": Object {
    "key": "back up",
  },
  "extendedData": undefined,
  "imageId": "9f85cf09-f254-4daf-8189-05b72a0fdb4d",
  "label": "Plain",
  "labelDefinition": Object {
    "displayValue": "Plain",
    "rawValue": "Plain",
    "typeName": "string",
  },
  "mergedFieldNames": Array [],
  "primaryKeys": Array [],
  "values": Object {
    "key": "Forward",
  },
}
`;

exports[`Item toJSON serializes item without label definition to json 1`] = `
Object {
  "classInfo": undefined,
  "displayValues": Object {
    "key": "models",
  },
  "extendedData": undefined,
  "imageId": "9cc4d132-a1f6-4c5c-85a2-222ae14a7e22",
  "label": Object {
    "displayValue": "Incredible",
    "rawValue": "Incredible",
    "typeName": "string",
  },
  "mergedFieldNames": Array [],
  "primaryKeys": Array [],
  "values": Object {
    "key": "solid state",
  },
}
`;
